### PR TITLE
[com_fields] Rename alias to name and add a check if name is unique

### PR DESCRIPTION
**Pull request by astridx**
last edit of this description 2017-01-16

Please have a look to this pr (https://github.com/joomla/joomla-cms/pull/13182) first. It has a solution for the same issue. But for the case that there is no agreement my solution would help solving the issue with the not unique alias. 

Pull Request for Issue # .
https://github.com/joomla/joomla-cms/issues/13335

### Summary of Changes
1. Changing the alias in name in database and code.
2. changed the position of the field -> position of field alias is empty
                                                       -> field name is shown with the options
3. add a check, if the field name is unique

### Testing Instructions
1. Change the value in the database for the com_fields alias to name

SET SQL_MODE = "NO_AUTO_VALUE_ON_ZERO";
SET time_zone = "+00:00";

ALTER TABLE `#__fields` CHANGE  `alias` `name` VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci NOT NULL DEFAULT '';

(or make new installation of Joomla after applying this path).

2. Apply this patch

3. Create and edit Custom fields in the frontend and in the backend 

### Documentation Changes Required
No

### CODE REVIEW

**Comment by laoneo** [comment]
Have planed to do a similar pr over the weekend, but you was faster. I would rename it to "name" only. As most joomla components do a check on store if the alias is already set (for example [here](https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_contact/tables/contact.php#L115) ), I would do the same with fields as well. Then you can remove the confusing uniqid from the name, it will look then much cleaner. Read only is then not necessary anymore as we allow the user to change the name then.

**Comment by astridx** [comment]
@laoneo Why do you want to make the alias/unique_name/name field changeable? If it is read-only you do not need to add the test for checking if it is unique. Is not that easier?

**Comment by laoneo** [comment]
So whats then the gain here at all, when you have a read only field with some cryptic id, IMO why not use then the id directly as suggested in #13576? For me must the form name be changeable, if it's not then we build another kind of id field.

The name stays even when you have changed the title so the admin will not be able to change the name as well? We will have the situation then that the title is not inline anymore with the name? I see here a big problem as well.

**Comment by astridx** [comment]
The gain is:
If i want to copy i custom layout to another web-site i will not have to remap the ids.

**Comment by laoneo** [comment]
You have to, because of the unique id which is always generated when the field is created. When we have a name the admin can define, then you can set up a field on the other site with the same name and copy the layout.

**Comment by Bakual** [comment]
Using `uniqid()` certainly doesn't make sense when we already have an unique ID that could be used.
Imho, we either use the ID as identifier like in my other PR or we rename the field to "name" and move it to the regular options (not after the title). During save we check that it is unique, if empty fall back to the ID. That would be the simplest approach

But I still don't see why that needs to be configurable. I doubt you're going to copy layouts from one site to another 1:1. You will have to adjust them anyway and it would be easy to adjust that ID.
Also, it is much more likely that the site admin will break that layout/design if it is a configurable field.

**Comment by astridx** [comment]
@laoneo  
For me the main problem is the alias, and that fields with the same alias only shown ones. So I thought the layout override problem not to the end.
You need a special field that you can manipulate for die override. Why is it not possible to use the title field for this override?

**Comment by astridx** [comment]
I changed the name from unique name to name.
The position of the field is now within the options.
I check if the name is unique.

File: administrator/language/en-GB/en-GB.ini [hunk 1]
```diff
@@ -177,9 +177,9 @@ JERROR_SAVE_FAILED="Could not save data. Error: %s"
 
 JFIELD_ACCESS_DESC="The access level group that is allowed to view this item."
 JFIELD_ACCESS_LABEL="Access"
-JFIELD_ALIAS_DESC="The Alias will be used in the SEF URL. Leave this blank and Joomla will fill in a default value from the title. This value will depend on the SEO settings (Global Configuration->Site). <br />Using Unicode will produce UTF-8 aliases. You may also enter manually any UTF-8 character. Spaces and some forbidden characters will be changed to hyphens.<br />When using default transliteration it will produce an alias in lower case and with dashes instead of spaces. You may enter the Alias manually. Use lowercase letters and hyphens (-). No spaces or underscores are allowed. Default value will be a date and time if the title is typed in non-latin letters ."
-JFIELD_ALIAS_LABEL="Alias"
-JFIELD_ALIAS_PLACEHOLDER="Auto-generate from title"
+JFIELD_NAME_DESC="The name will be used for identifying the field. Joomla! will fill in a value computed from the title."
```
Comment (laoneo) [inline administrator/language/en-GB/en-GB.ini]: You can't rename that string as this string is used by other components which have an alias. You need to create a new ine un com_fields.

**Comment by laoneo** [comment]
The MSSQL files nees to be adapted as well.

File: administrator/language/en-GB/en-GB.ini (hunk 1 above)
Comment (astridx) [inline administrator/language/en-GB/en-GB.ini]: Are you sure? Where is this language string used?I searched the whole projekt, but I only found it in JFields and the prefix indicates this too, doen't it?

Comment (infograf768) [inline administrator/language/en-GB/en-GB.ini]: screen shot 2017-01-15 at 10 34 40

**Comment by jeckodevelopment** [comment]
@astridx please look at the conflicts :)

**Comment by astridx** [comment]
This was my first conflict editing. I hope everything is correct :)

**Comment by laoneo** [comment]
I like the PR very much and it is done the way it needs to be now. You need to change the description and title of this pr as it behaves now differently with the changes you made in the last few commits.

I'v tested it and it works as expected. There is now a missing language string
image
Fater this one gets fixed, I can mark the test as successful. Good work!

**Comment by astridx** [comment]
I'm so sorry. I had forgotten this text when renaming.
Thanks for the patience with me and the tips :)

**Comment by astridx** [comment]
I'm so sorry. I had forgotten this text when renaming.
Thanks for the patience with me and the tips :)

**Comment by laoneo** [comment]
I have tested this item :white_check_mark: successfully on 1f3d9913b03e41e5405978ce4327bcb2ae81e5bc<br /><br />No need to apologize, we need people who are willing to help.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13576">issues.joomla.org/tracker/joomla-cms/13576</a>.</sub>

**Comment by laoneo** [comment]
It would be good to change the pr description as well as it does now something different.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/13576">issues.joomla.org/tracker/joomla-cms/13576</a>.</sub>

**Comment by Bakual** [comment]
@franz-wohlkoenig found another issue with allowing to set the name of the field: https://github.com/joomla/joomla-cms/pull/13621#issuecomment-273881546

Basically you can overwrite existing fields of the original form when you name it the same as the existing field. At least as long as we store the fields in the "params" group of the form which may be used already by the original form.

**Comment by astridx** [comment]
Would it be a compromise to take this PR in combination with  https://github.com/joomla/joomla-cms/pull/13353 ?
So the people who need the name field could use it and we solved the problem with overwrite existing fields.

**Comment by Bakual** [comment]
Yep, that would solve the overwrite issue. But it's not about being a compromise, that one needs to go in anyway 😄

**Comment by laoneo** [comment]
Please solve the merge conflicts, that we can get this one in.

**Comment by astridx** [comment]
Merge conflicts are because of https://github.com/joomla/joomla-cms/pull/13353 and I solved them now.

**Comment by laoneo** [comment]
Thanks

**Comment by Bakual** [comment]
Can you fix the conflicts please?

**Comment by astridx** [comment]
I just did it.

**Comment by astridx** [comment]
Do I have to do something because of the not successful test "continuous-integration/appveyor/pr — AppVeyor build failed" ?

**Comment by laoneo** [comment]
No, almost any pr fails at the moment.

**Comment by laoneo** [comment]
Can you again please fix the conflicts please. Then we should get that one in before the first RC. @Bakual or @rdeutz can this being merged then?

**Comment by astridx** [comment]
At some points I am not sure what I have to delete and what is correct. I think it would be easier for me to create a new pr. Should i do this?

**Comment by rdeutz** [comment]
@astridx how ever it works best for you it needs new testing anyway.

**Comment by astridx** [comment]
Closed because of new PR https://github.com/joomla/joomla-cms/pull/14554

**Pull request closed without merging**